// bunabingo-backend/src/bot/commands/admin.ts
import { Context, Markup } from 'telegraf';
import { isAdmin } from '../../services/user.service';
import { getPendingDeposits, approveDeposit, rejectDeposit } from '../../services/deposit.service';
import { getPendingWithdrawals, approveWithdrawal, rejectWithdrawal } from '../../services/withdrawal.service';
import { getUserByTelegramId } from '../../services/user.service';
import prisma from '../../lib/prisma';
import { config } from '../../config';

async function requireAdmin(ctx: Context): Promise<string | null> {
  const tgUser = ctx.from!;
  if (!isAdmin(tgUser.id)) {
    await ctx.reply('тЭМ Admin only command.');
    return null;
  }
  const user = await getUserByTelegramId(tgUser.id);
  return user?.id ?? null;
}

export async function handleAdminPanel(ctx: Context) {
  const tgUser = ctx.from!;
  if (!isAdmin(tgUser.id)) return ctx.reply('тЭМ Unauthorized.');

  const [pendingDeposits, pendingWithdrawals, totalUsers, activeGames] = await Promise.all([
    prisma.deposit.count({ where: { status: 'PENDING' } }),
    prisma.withdrawal.count({ where: { status: 'PENDING' } }),
    prisma.user.count(),
    prisma.game.count({ where: { status: { in: ['RUNNING', 'COUNTDOWN', 'WAITING'] } } }),
  ]);

  await ctx.reply(
    `ЁЯЫбя╕П *Admin Panel*\n\n` +
    `ЁЯУе Pending Deposits: *${pendingDeposits}*\n` +
    `ЁЯУд Pending Withdrawals: *${pendingWithdrawals}*\n` +
    `ЁЯСе Total Users: *${totalUsers}*\n` +
    `ЁЯОо Active Games: *${activeGames}*\n`,
    {
      parse_mode: 'Markdown',
      ...Markup.inlineKeyboard([
        [
          Markup.button.callback(`ЁЯУе Deposits (${pendingDeposits})`, 'admin_deposits'),
          Markup.button.callback(`ЁЯУд Withdrawals (${pendingWithdrawals})`, 'admin_withdrawals'),
        ],
        [Markup.button.webApp('ЁЯУК Full Dashboard', `${config.bot.miniAppUrl}/admin`)],
      ]),
    }
  );
}

export async function handleAdminDeposits(ctx: Context) {
  const adminId = await requireAdmin(ctx);
  if (!adminId) return;

  const deposits = await getPendingDeposits();
  if (!deposits.length) {
    return ctx.reply('тЬЕ No pending deposits.');
  }

  for (const dep of deposits.slice(0, 5)) {
    const userName = dep.user.telegramUsername ? `@${dep.user.telegramUsername}` : dep.user.firstName;
    await ctx.reply(
      `ЁЯУе *Deposit Request*\n\n` +
      `ЁЯСд User: ${userName}\n` +
      `ЁЯТ╡ Amount: *${Number(dep.amount).toFixed(2)} ETB*\n` +
      `ЁЯФЦ Reference: ${dep.reference || 'N/A'}\n` +
      `ЁЯУЕ Submitted: ${dep.createdAt.toLocaleString()}`,
      {
        parse_mode: 'Markdown',
        ...Markup.inlineKeyboard([
          [
            Markup.button.callback('тЬЕ Approve', `approve_dep_${dep.id}`),
            Markup.button.callback('тЭМ Reject', `reject_dep_${dep.id}`),
          ],
        ]),
      }
    );
  }
}

export async function handleAdminWithdrawals(ctx: Context) {
  const adminId = await requireAdmin(ctx);
  if (!adminId) return;

  const withdrawals = await getPendingWithdrawals();
  if (!withdrawals.length) {
    return ctx.reply('тЬЕ No pending withdrawals.');
  }

  for (const wd of withdrawals.slice(0, 5)) {
    const userName = wd.user.telegramUsername ? `@${wd.user.telegramUsername}` : wd.user.firstName;
    await ctx.reply(
      `ЁЯУд *Withdrawal Request*\n\n` +
      `ЁЯСд User: ${userName}\n` +
      `ЁЯТ╡ Amount: *${Number(wd.amount).toFixed(2)} ETB*\n` +
      `ЁЯПж Bank: ${wd.bankName}\n` +
      `ЁЯСд Account: ${wd.accountName}\n` +
      `ЁЯФв Account #: ${wd.accountNumber}\n` +
      `ЁЯУЕ Submitted: ${wd.createdAt.toLocaleString()}`,
      {
        parse_mode: 'Markdown',
        ...Markup.inlineKeyboard([
          [
            Markup.button.callback('тЬЕ Approve', `approve_wd_${wd.id}`),
            Markup.button.callback('тЭМ Reject', `reject_wd_${wd.id}`),
          ],
        ]),
      }
    );
  }
}

export async function handleApproveDeposit(ctx: Context, depositId: string) {
  const adminId = await requireAdmin(ctx);
  if (!adminId) return;
  try {
    await approveDeposit(depositId, adminId);
    await ctx.answerCbQuery('тЬЕ Deposit approved!');
    await ctx.editMessageText(`тЬЕ *Deposit Approved*\nID: \`${depositId}\``, { parse_mode: 'Markdown' });
  } catch (e: any) {
    await ctx.answerCbQuery(`тЭМ ${e.message}`);
  }
}

export async function handleRejectDeposit(ctx: Context, depositId: string) {
  const adminId = await requireAdmin(ctx);
  if (!adminId) return;
  try {
    await rejectDeposit(depositId, adminId, 'Rejected by admin');
    await ctx.answerCbQuery('тЭМ Deposit rejected');
    await ctx.editMessageText(`тЭМ *Deposit Rejected*\nID: \`${depositId}\``, { parse_mode: 'Markdown' });
  } catch (e: any) {
    await ctx.answerCbQuery(`тЭМ ${e.message}`);
  }
}

export async function handleApproveWithdrawal(ctx: Context, withdrawalId: string) {
  const adminId = await requireAdmin(ctx);
  if (!adminId) return;
  try {
    await approveWithdrawal(withdrawalId, adminId);
    await ctx.answerCbQuery('тЬЕ Withdrawal approved!');
    await ctx.editMessageText(`тЬЕ *Withdrawal Approved*\nID: \`${withdrawalId}\``, { parse_mode: 'Markdown' });
  } catch (e: any) {
    await ctx.answerCbQuery(`тЭМ ${e.message}`);
  }
}

export async function handleRejectWithdrawal(ctx: Context, withdrawalId: string) {
  const adminId = await requireAdmin(ctx);
  if (!adminId) return;
  try {
    await rejectWithdrawal(withdrawalId, adminId, 'Rejected by admin');
    await ctx.answerCbQuery('тЭМ Withdrawal rejected');
    await ctx.editMessageText(`тЭМ *Withdrawal Rejected*\nID: \`${withdrawalId}\``, { parse_mode: 'Markdown' });
  } catch (e: any) {
    await ctx.answerCbQuery(`тЭМ ${e.message}`);
  }
}
