// bunabingo-backend/dist/bot/commands/depositFlow.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleDepositManualStart = handleDepositManualStart;
exports.handleDepositCancel = handleDepositCancel;
exports.handleDepositSubmit = handleDepositSubmit;
exports.handleDepositMessage = handleDepositMessage;
exports.handlePayCbeBirr = handlePayCbeBirr;
exports.handlePayCbeBank = handlePayCbeBank;
exports.handlePayMpesa = handlePayMpesa;
exports.handlePayTelebirr = handlePayTelebirr;
const telegraf_1 = require("telegraf");
const user_service_1 = require("../../services/user.service");
const session_1 = require("../session");
const config_1 = require("../../config");
const prisma_1 = __importDefault(require("../../lib/prisma"));
const logger_1 = require("../../lib/logger");
// в”Җв”Җв”Җ Helpers в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
function generateReference(length = 10) {
    const chars = 'ABCDEFGHJKLMNPQRSTUVWXYZabcdefghjkmnpqrstuvwxyz23456789';
    return Array.from({ length }, () => chars[Math.floor(Math.random() * chars.length)]).join('');
}
const CANCEL_BTN = [[telegraf_1.Markup.button.callback('вқҢ Cancel', 'cmd_deposit_cancel')]];
// в”Җв”Җв”Җ Step 1: Ask for amount в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
async function handleDepositManualStart(ctx) {
    if (ctx.callbackQuery)
        await ctx.answerCbQuery();
    const tgUser = ctx.from;
    const user = await (0, user_service_1.getUserByTelegramId)(tgUser.id);
    if (!user)
        return ctx.reply('вқҢ Please /start first to register.');
    (0, session_1.setSession)(tgUser.id, { type: 'MANUAL_DEPOSIT', step: 'AWAITING_AMOUNT' });
    await ctx.reply(`рҹ’і *Manual Deposit*\n\n` +
        `бҠҘбҠ•бӢІбҲһбҲӢбҲҚбӢҺбүө бӢЁбҲҡбҚҲбҲҚбҢүбүөбҠ• бӢЁбҢҲбҠ•бӢҳбүҘ бҲҳбҢ бҠ• бӢ«бҲөбҢҲбүЎ:\n` +
        `_(Enter the amount you want to deposit in ETB)_\n\n` +
        `_Minimum: 10 ETB_`, {
        parse_mode: 'Markdown',
        ...telegraf_1.Markup.inlineKeyboard(CANCEL_BTN),
    });
}
// в”Җв”Җв”Җ Cancel в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
async function handleDepositCancel(ctx) {
    if (ctx.callbackQuery)
        await ctx.answerCbQuery();
    (0, session_1.clearSession)(ctx.from.id);
    await ctx.reply('вқҢ Deposit cancelled.');
}
// в”Җв”Җв”Җ Skip screenshot вҶ’ submit в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
async function handleDepositSubmit(ctx) {
    if (ctx.callbackQuery)
        await ctx.answerCbQuery();
    const session = (0, session_1.getSession)(ctx.from.id);
    if (!session || session.type !== 'MANUAL_DEPOSIT')
        return;
    await submitDeposit(ctx, session.amount, session.reference, undefined, session.paymentMethod);
}
// в”Җв”Җв”Җ Main message router в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
async function handleDepositMessage(ctx) {
    const tgUser = ctx.from;
    const session = (0, session_1.getSession)(tgUser.id);
    if (!session || session.type !== 'MANUAL_DEPOSIT')
        return false;
    const msg = ctx.message; // Cast to any to handle both text and photo properties safely
    // в”Җв”Җ AWAITING_AMOUNT в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
    if (session.step === 'AWAITING_AMOUNT') {
        const raw = msg?.text?.trim();
        const amount = parseFloat(raw);
        if (!raw || isNaN(amount) || amount < 10) {
            await ctx.reply(`вҡ пёҸ Invalid amount. Please enter a number вүҘ 10 ETB.\n\nбҠҘбүЈбҠӯбӢҺ бүөбҠӯбҠӯбҲҲбҠӣ бӢЁбҢҲбҠ•бӢҳбүҘ бҲҳбҢ бҠ• бӢ«бҲөбҢҲбүЎ:`, { parse_mode: 'Markdown', ...telegraf_1.Markup.inlineKeyboard(CANCEL_BTN) });
            return true;
        }
        const reference = generateReference();
        (0, session_1.setSession)(tgUser.id, {
            type: 'MANUAL_DEPOSIT',
            step: 'AWAITING_SCREENSHOT',
            amount,
            reference,
        });
        const { receiverName, receiverPhone } = config_1.config.payment;
        await ctx.reply(`Payment details`, { parse_mode: 'Markdown' });
        await ctx.reply(`\`\`\`\nName:      ${receiverName}\nPhone:     ${receiverPhone}\nAmount:    ${amount}ETB\nreference: ${reference}\n\`\`\``, {
            parse_mode: 'Markdown',
            ...telegraf_1.Markup.inlineKeyboard([
                [telegraf_1.Markup.button.callback('Pay from  CBE-Birr to MPESA', 'cmd_pay_cbe_birr')],
                [telegraf_1.Markup.button.callback('Pay from  CBEBANK to MPESA', 'cmd_pay_cbe_bank')],
                [telegraf_1.Markup.button.callback('Pay from  MPESA to MPESA only', 'cmd_pay_mpesa')],
                [telegraf_1.Markup.button.callback('Pay from  telebirr to telebirr only', 'cmd_pay_telebirr')],
                [telegraf_1.Markup.button.callback('вқҢ Cancel', 'cmd_deposit_cancel')],
            ]),
        });
        return true;
    }
    // в”Җв”Җ AWAITING_SCREENSHOT (CBE/MPESA) в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
    if (session.step === 'AWAITING_SCREENSHOT') {
        const photoMsg = msg;
        if (!photoMsg.photo?.length) {
            await ctx.reply(`рҹ“ё Please send your payment *screenshot*, or click Skip.\n_(бӢ«бҲҲбӢҺбүөбҠ• бӢ°бҲЁбҲ°бҠқ бҚҺбү¶ бӢӯбҲӢбҠ©)_`, {
                parse_mode: 'Markdown',
                ...telegraf_1.Markup.inlineKeyboard([
                    [telegraf_1.Markup.button.callback('вҸӯ Skip Screenshot', 'cmd_deposit_submit')],
                    [telegraf_1.Markup.button.callback('вқҢ Cancel', 'cmd_deposit_cancel')],
                ]),
            });
            return true;
        }
        const fileId = photoMsg.photo[photoMsg.photo.length - 1].file_id;
        // --- Automatic OCR Validation for Photos ---
        await ctx.reply('рҹ”Қ Scanning receipt for automatic approval...');
        const file = await ctx.telegram.getFile(fileId);
        const fileUrl = `https://api.telegram.org/file/bot${config_1.config.bot.token}/${file.file_path}`;
        const { validateReceiptImage } = await Promise.resolve().then(() => __importStar(require('../../services/bunafrankValidator')));
        const result = await validateReceiptImage(fileUrl, session.amount);
        if (result.valid) {
            const d = result.data;
            await ctx.reply(`вң… *Automatic Verification Successful!*\nTransaction ID: \`${d.transactionId}\`\nAmount: \`${d.amount} ETB\``, { parse_mode: 'Markdown' });
            await submitDeposit(ctx, session.amount, d.transactionId, fileId, session.paymentMethod, true);
        }
        else {
            await ctx.reply(`вҡ пёҸ *Manual Review Required*\nReason: ${result.error}\n\nSubmitting for admin review...`, { parse_mode: 'Markdown' });
            await submitDeposit(ctx, session.amount, session.reference, fileId, session.paymentMethod, false, result.error);
        }
        return true;
    }
    // в”Җв”Җ AWAITING_SMS (Telebirr) в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
    if (session.step === 'AWAITING_SMS') {
        const smsText = msg?.text?.trim();
        if (!smsText || smsText.length < 20) {
            await ctx.reply(`вҡ пёҸ Please paste the *full* Telebirr SMS you received.\n_(Starting with "Dear...")_`, { parse_mode: 'Markdown', ...telegraf_1.Markup.inlineKeyboard(CANCEL_BTN) });
            return true;
        }
        // в”Җв”Җ Validate в”Җв”Җ
        await ctx.reply(`рҹ”Қ Validating your Telebirr receipt...`);
        const { validateTelebirrSms, validateReceiptImage } = await Promise.resolve().then(() => __importStar(require('../../services/bunafrankValidator')));
        const result = await validateTelebirrSms(smsText, session.amount, config_1.config.payment.telebirrPhone);
        if (!result.valid) {
            await ctx.reply(result.error + `\n\nPlease try again or contact support.`, { parse_mode: 'Markdown', ...telegraf_1.Markup.inlineKeyboard(CANCEL_BTN) });
            return true;
        }
        // в”Җв”Җ Valid вҖ” show parsed confirmation в”Җв”Җ
        const d = result.data;
        const verifiedBadge = result.onlineVerified ? 'вң… Verified online' : 'вҡ пёҸ Pending manual review';
        await ctx.reply(`вң… *Receipt Validated!*\n\n` +
            `\`\`\`\n` +
            `Transaction ID : ${d.transactionId}\n` +
            `Amount         : ETB ${d.amount.toFixed(2)}\n` +
            `Recipient      : ${d.recipientName}\n` +
            `Phone          : ${d.recipientPhoneMasked}\n` +
            `Date/Time      : ${d.dateTime}\n` +
            `Service Fee    : ETB ${d.serviceFee.toFixed(2)}\n` +
            `\`\`\`\n` +
            `рҹ”— ${verifiedBadge}\n\n` +
            `Submitting your deposit...`, { parse_mode: 'Markdown' });
        // Use transaction ID as the reference (unique & verifiable)
        await submitDeposit(ctx, session.amount, d.transactionId, undefined, 'telebirr');
        return true;
    }
    return false;
}
// в”Җв”Җв”Җ Payment method: CBE-Birr в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
async function handlePayCbeBirr(ctx) {
    if (ctx.callbackQuery)
        await ctx.answerCbQuery();
    const session = (0, session_1.getSession)(ctx.from.id);
    if (!session || session.type !== 'MANUAL_DEPOSIT')
        return;
    (0, session_1.setSession)(ctx.from.id, { ...session, paymentMethod: 'cbe_birr', step: 'AWAITING_SCREENSHOT' });
    const { receiverPhone } = config_1.config.payment;
    await ctx.reply(`рҹҸҰ *CBE-Birr вҶ’ MPESA*\n\n` +
        `\`\`\`\n` +
        `1. CBE-Birr бҠ бҚ•бҲҚбҠ¬бҲҪбҠ• бӢӯбҠӯбҚҲбүұ\n` +
        `2. ${session.amount} бүҘбҲӯ бӢҲбӢ° ${receiverPhone} бӢӯбҲӢбҠ©\n` +
        `3. бӢ°бҲЁбҲ°бҠқ (screenshot) бӢ«бҲөбүҖбҲқбҢЎ\n` +
        `4. бӢ«бҲөбүҖбҲҳбҢЎбүөбҠ• бҚҺбү¶ бҠЁбӢҡ бҲӢбӢӯ бӢӯбҲӢбҠ©\n` +
        `\`\`\`\n\n` +
        `рҹ“ё Send your *payment screenshot* here рҹ‘Ү`, {
        parse_mode: 'Markdown',
        ...telegraf_1.Markup.inlineKeyboard([
            [telegraf_1.Markup.button.callback('вҸӯ Skip Screenshot', 'cmd_deposit_submit')],
            [telegraf_1.Markup.button.callback('вқҢ Cancel', 'cmd_deposit_cancel')],
        ]),
    });
}
// в”Җв”Җв”Җ Payment method: CBE Bank в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
async function handlePayCbeBank(ctx) {
    if (ctx.callbackQuery)
        await ctx.answerCbQuery();
    const session = (0, session_1.getSession)(ctx.from.id);
    if (!session || session.type !== 'MANUAL_DEPOSIT')
        return;
    (0, session_1.setSession)(ctx.from.id, { ...session, paymentMethod: 'cbe_bank', step: 'AWAITING_SCREENSHOT' });
    const { receiverPhone } = config_1.config.payment;
    await ctx.reply(`рҹҸҰ *CBE Bank вҶ’ MPESA*\n\n` +
        `\`\`\`\n` +
        `1. CBEBirr бӢҲбӢӯбҲқ бү…бҲӯбүҘ бӢҲбӢ°бҲҶбҠҗ CBE бү…бҲӯбҠ•бҢ«бҚҚ бӢӯбҲӮбӢұ\n` +
        `2. ${session.amount} бүҘбҲӯ бӢҲбӢ° ${receiverPhone} бӢ«бҲөбү°бҲӢбҲҚбҚү\n` +
        `3. бӢ°бҲЁбҲ°бҠқ (screenshot) бӢ«бҲөбүҖбҲқбҢЎ\n` +
        `4. бӢ«бҲөбүҖбҲҳбҢЎбүөбҠ• бҚҺбү¶ бҠЁбӢҡ бҲӢбӢӯ бӢӯбҲӢбҠ©\n` +
        `\`\`\`\n\n` +
        `рҹ“ё Send your *payment screenshot* here рҹ‘Ү`, {
        parse_mode: 'Markdown',
        ...telegraf_1.Markup.inlineKeyboard([
            [telegraf_1.Markup.button.callback('вҸӯ Skip Screenshot', 'cmd_deposit_submit')],
            [telegraf_1.Markup.button.callback('вқҢ Cancel', 'cmd_deposit_cancel')],
        ]),
    });
}
// в”Җв”Җв”Җ Payment method: MPESA в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
async function handlePayMpesa(ctx) {
    if (ctx.callbackQuery)
        await ctx.answerCbQuery();
    const session = (0, session_1.getSession)(ctx.from.id);
    if (!session || session.type !== 'MANUAL_DEPOSIT')
        return;
    (0, session_1.setSession)(ctx.from.id, { ...session, paymentMethod: 'mpesa', step: 'AWAITING_SCREENSHOT' });
    const { receiverPhone } = config_1.config.payment;
    await ctx.reply(`рҹ“ұ *MPESA вҶ’ MPESA*\n\n` +
        `\`\`\`\n` +
        `1. MPESA бҠ бҚ•бҲҚбҠ¬бҲҪбҠ• бӢӯбҠӯбҚҲбүұ\n` +
        `2. ${session.amount} бүҘбҲӯ бӢҲбӢ° ${receiverPhone} бӢӯбҲӢбҠ©\n` +
        `3. бӢ°бҲЁбҲ°бҠқ (screenshot) бӢ«бҲөбүҖбҲқбҢЎ\n` +
        `4. бӢ«бҲөбүҖбҲҳбҢЎбүөбҠ• бҚҺбү¶ бҠЁбӢҡ бҲӢбӢӯ бӢӯбҲӢбҠ©\n` +
        `\`\`\`\n\n` +
        `рҹ“ё Send your *payment screenshot* here рҹ‘Ү`, {
        parse_mode: 'Markdown',
        ...telegraf_1.Markup.inlineKeyboard([
            [telegraf_1.Markup.button.callback('вҸӯ Skip Screenshot', 'cmd_deposit_submit')],
            [telegraf_1.Markup.button.callback('вқҢ Cancel', 'cmd_deposit_cancel')],
        ]),
    });
}
// в”Җв”Җв”Җ Payment method: Telebirr в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
async function handlePayTelebirr(ctx) {
    if (ctx.callbackQuery)
        await ctx.answerCbQuery();
    const session = (0, session_1.getSession)(ctx.from.id);
    if (!session || session.type !== 'MANUAL_DEPOSIT')
        return;
    (0, session_1.setSession)(ctx.from.id, { ...session, paymentMethod: 'telebirr', step: 'AWAITING_SMS' });
    const { telebirrPhone, supportAgent1, supportAgent2 } = config_1.config.payment;
    await ctx.reply(`бӢЁбүҙбҲҢбүҘбҲӯ бҠ бҠ«бӢҚбҠ•бүө\n\n` +
        `\`${telebirrPhone}\`\n\n` +
        `\`\`\`\n` +
        `1. бүҙбҲҢбүҘбҲӯ бҠЁ ${telebirrPhone} бҲӢбӢӯ ${session.amount} бүҘбҲӯ бӢӯбҲӢбҠ©\n` +
        `2. бҠЁбүҙбҲҢбүҘбҲӯ бӢ«бҢҲбҠҷбүөбҠ• SMS бӢӯбҢ бүҘбүҒ\n` +
        `3. SMS бҲІбӢ°бҲӯбҲөбӢҺ бӢ«бҠ•бҠ‘ бӢ«бҲөбүҖбҲқбҢЎ (copy)\n` +
        `4. бӢ«бҲөбүҖбҲҳбҢЎбүөбҠ• (sms) бҢҪбҲҒбҚҚ бҠЁ(copy) бҠ бӢөбҲӯбҢү бҠЁбӢҡ бҲӢбӢӯ бӢӯбҲҲбҢҘбҚү(paste)\n` +
        `\`\`\`\n\n` +
        `бӢЁбҲҡбӢ«бҢӢбҢҘбҲӣбү№ бӢЁбҠӯбҚҚбӢ« бүҪбҢҚбҲӯ бҠ«бҲҲ ${supportAgent1} бү бӢҡ бҠӨбҢҖбҠ•бүөбҠ• бҲӣбӢӢбҲ«бүө бӢӯбүҪбҲӢбҲү бӢҲбӢӯбҲқ ${supportAgent2} бү бӢҡ бҲібҚ–бҲӯбүө бҲӣбӢүбҲ«бүө бӢӯбүҪбҲӢбҲү\n\n` +
        `бӢЁбҠЁбҚҲбҲҲбүҪбҲҒбү бүөбҠ• бҠ бҢӯбҲӯ бӢЁбҢ№бҲҒбҚҚ бҲҳбҲҲбҠӯбүө(sms) бҠҘбӢҡ бҲӢбӢӯ бӢ«бҲөбҢҲбүЎбүө рҹ‘Үрҹ‘Үрҹ‘Ү`, {
        parse_mode: 'Markdown',
        ...telegraf_1.Markup.inlineKeyboard(CANCEL_BTN),
    });
}
// в”Җв”Җв”Җ Final submission в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җ
async function submitDeposit(ctx, amount, referenceOrSms, screenshotFileId, paymentMethod, autoApprove = false, ocrFailureReason) {
    const tgUser = ctx.from;
    (0, session_1.clearSession)(tgUser.id);
    try {
        const user = await (0, user_service_1.getUserByTelegramId)(tgUser.id);
        if (!user)
            return ctx.reply('вқҢ User not found. Please /start first.');
        const deposit = await prisma_1.default.deposit.create({
            data: {
                userId: user.id,
                amount,
                reference: referenceOrSms,
                receiptUrl: screenshotFileId ?? null,
                status: 'PENDING',
            },
        });
        logger_1.logger.info(`[Deposit] ${deposit.id} вҖ” ${amount} ETB вҖ” method: ${paymentMethod ?? 'unknown'}`);
        const methodLabel = paymentMethod === 'telebirr' ? 'Telebirr' :
            paymentMethod === 'cbe_birr' ? 'CBE-Birr' :
                paymentMethod === 'cbe_bank' ? 'CBE Bank' :
                    paymentMethod === 'mpesa' ? 'MPESA' : 'Manual';
        const isTelebirr = paymentMethod === 'telebirr' || (referenceOrSms && /^[A-Z0-9]{10}$/.test(referenceOrSms));
        const receiptUrl = isTelebirr ? `https://transactioninfo.ethiotelecom.et/receipt/${referenceOrSms}` : null;
        await ctx.reply(`вң… *Deposit Submitted Successfully!*\n\n` +
            `рҹ’ө Amount: *${amount.toFixed(2)} ETB*\n` +
            `рҹ’і Method: *${methodLabel}*\n` +
            `рҹ“Ӣ Status: *${autoApprove ? 'вң… Approved' : 'Pending Review'}*\n\n` +
            (autoApprove
                ? `Your balance has been updated! вҳ•`
                : `вҸұ Your deposit will be reviewed within *30 minutes*.\nYou will be notified once approved. рҹҷҸ`), { parse_mode: 'Markdown' });
        if (autoApprove) {
            const { approveDeposit } = await Promise.resolve().then(() => __importStar(require('../../services/deposit.service')));
            const systemAdminId = config_1.config.bot.adminIds[0] || 'SYSTEM';
            try {
                await approveDeposit(deposit.id, systemAdminId);
                logger_1.logger.info(`[Deposit] Auto-approved ${deposit.id}`);
            }
            catch (err) {
                logger_1.logger.error(`[Deposit] Auto-approve failed for ${deposit.id}:`, err);
            }
            return; // No need to notify admins for manual approval if already done
        }
        // Notify admins
        const userName = tgUser.username ? `@${tgUser.username}` : user.firstName;
        const isSms = paymentMethod === 'telebirr';
        const adminCaption = `рҹ“Ҙ *New Manual Deposit вҖ” ${methodLabel}*\n\n` +
            `рҹ‘Ө User: ${userName}\n` +
            `рҹ’ө Amount: *${amount.toFixed(2)} ETB*\n` +
            (ocrFailureReason ? `вҡ пёҸ *OCR FAIL:* \`${ocrFailureReason}\`\n` : '') +
            (isSms
                ? `рҹ“ұ SMS Receipt:\n\`\`\`\n${referenceOrSms}\n\`\`\``
                : `рҹ”– Reference: \`${referenceOrSms}\``) +
            (receiptUrl ? `\nрҹ”— [View Official Receipt](${receiptUrl})` : '') +
            `\nрҹҶ” Deposit ID: \`${deposit.id}\``;
        const adminKeyboard = telegraf_1.Markup.inlineKeyboard([
            [
                telegraf_1.Markup.button.callback('вң… Approve', `approve_dep_${deposit.id}`),
                telegraf_1.Markup.button.callback('вқҢ Reject', `reject_dep_${deposit.id}`),
            ],
        ]);
        for (const adminIdStr of config_1.config.bot.adminIds) {
            try {
                const adminTgId = parseInt(adminIdStr, 10);
                if (screenshotFileId) {
                    await ctx.telegram.sendPhoto(adminTgId, screenshotFileId, {
                        caption: adminCaption,
                        parse_mode: 'Markdown',
                        ...adminKeyboard,
                    });
                }
                else {
                    await ctx.telegram.sendMessage(adminTgId, adminCaption, {
                        parse_mode: 'Markdown',
                        ...adminKeyboard,
                    });
                }
            }
            catch (e) {
                logger_1.logger.warn(`[Deposit] Could not notify admin ${adminIdStr}:`, e);
            }
        }
        // --- Optional: Notify Support Admin specifically for Manual Reviews ---
        if (!autoApprove && config_1.config.payment.supportAdminId) {
            try {
                const supportId = parseInt(config_1.config.payment.supportAdminId, 10);
                if (screenshotFileId) {
                    await ctx.telegram.sendPhoto(supportId, screenshotFileId, {
                        caption: `рҹҡЁ *Support Alert: Manual Review Needed*\n\n${adminCaption}`,
                        parse_mode: 'Markdown',
                        ...adminKeyboard,
                    });
                }
                else {
                    await ctx.telegram.sendMessage(supportId, `рҹҡЁ *Support Alert: Manual Review Needed*\n\n${adminCaption}`, {
                        parse_mode: 'Markdown',
                        ...adminKeyboard,
                    });
                }
            }
            catch (e) {
                logger_1.logger.warn(`[Deposit] Could not notify Support Admin:`, e);
            }
        }
    }
    catch (err) {
        logger_1.logger.error('[Deposit] Submit error:', err);
        await ctx.reply('вқҢ Something went wrong. Please try again or contact support.');
    }
}
